Домашнее задание к спринту 6

baaazik opened the pull request:
<img width="502" alt="image" src="https://github.com/user-attachments/assets/4d0a49f2-10fe-4f8c-92ea-53f175aedba8" />

baaazik committed:
feature: implement movies loading from IMDB

baaazik committed:
feature: use loaded movies in quiz

baaazik committed:
fix: fix loading error handling

IgorChumikov commented on MovieQuiz/Services/NetworkClient.swift:
```diff
@@ -0,0 +1,37 @@
+            }
+
+            if let response = response as? HTTPURLResponse,
+                response.statusCode < 200 || response.statusCode >= 300 {
```
Можно так 
```
guard let httpResponse = response as? HTTPURLResponse else {
    ...
}
guard (200..<300).contains(httpResponse.statusCode) else {
    ...
}
```
Почему лучше: guard делает поток выполнения более линейным и читаемым — меньше вложенности, проще поддерживать.

baaazik replied:
Сделано

IgorChumikov commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -8,7 +8,7 @@
 import Foundation
 
 class QuestionFactory: QuestionFactoryProtocol {
```
final

baaazik replied:
Сделано

IgorChumikov commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -8,7 +8,7 @@
 class QuestionFactory: QuestionFactoryProtocol {
-    private let questions: [QuizQuestion] = [
+    /* private let questions: [QuizQuestion] = [
         QuizQuestion(image: "The Godfather", text: "Рейтинг этого фильма больше чем 6?", correctAnswer: true),
```
удалить)

baaazik replied:
Сделано

IgorChumikov commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -19,22 +19,63 @@ class QuestionFactory: QuestionFactoryProtocol {
+    init(delegate: QuestionFactoryDelegate, moviesLoader: MoviesLoading) {
         self.delegate = delegate
-        self.currentRoundQuestions = questions.shuffled()
+        self.currentRoundMovies = movies.shuffled()
```
На момент инициализации movies ещё пустой, поэтому movies.shuffled() просто возвращает []. Это лишняя операция. Лучше сразу задать пустой массив, а наполнение логично оставить для loadData.

baaazik replied:
Сделано

IgorChumikov reviewed (commented):
Код написан очень достойно — видно, что ты хорошо разбираешься в архитектуре и потоках данных 💪
Использование делегатов и асинхронных вызовов сделано правильно и по всем канонам iOS-разработки 👏
Методы логично разделены, всё читается легко и понятно — это уже уровень🔥
Очень радует, что соблюдается инкапсуляция и слабые ссылки (weak) — это прямо отлично! ✅
Продолжай в том же духе — ты на правильном пути, и уже кодишь лучше, чем многие! 🚀

baaazik committed:
fix: minor fixes

- add final to classes
- remove commented out code
- remove unused shuffle
- make error handling in NetworkClient a bit cleaner

baaazik committed:
refactor: improve error handling

- handle IMDB API error message

baaazik committed:
fix: handle image loading errors

baaazik committed:
feature: use median rating to ask fair questions
